src/backtracking/hamiltonian_cycle.rs

```diff
@@ -0,0 +1,238 @@
+    /// This function is called by `find_hamiltonian_cycle`.
+    ///
+    /// # Arguments
+    ///
+    /// * `path` - A mutable vector representing the current path being explored.
+    /// * `pos` - The position of the current vertex being considered.
+    ///
+    /// # Returns
+    ///
+    /// * `true` if a Hamiltonian cycle is found, `false` otherwise.
+    fn hamiltonian_cycle_util(&self, path: &mut Vec<usize>, pos: usize) -> bool {
+        let vertices = self.adjacency_matrix.len();
+        if pos == vertices {
+            // Check if there is an edge from the last included vertex to the first vertex
+            return self.adjacency_matrix[path[pos - 1]][path[0]];
+        }
+
+        for v in 0..vertices {
+            if self.is_safe(v, path, pos) {
+                path[pos] = v;
+                if self.hamiltonian_cycle_util(path, pos + 1) {
+                    return true;
+                }
+                path[pos] = usize::MAX;
+            }
+        }
+
+        false
+    }
+
+    /// Finds a Hamiltonian cycle in the graph, if one exists, starting from the specified vertex.
+    ///
+    /// A Hamiltonian cycle is a cycle that visits every vertex exactly once
+    /// and returns to the starting vertex.
+    ///
+    /// # Note
+    /// This implementation may not find all possible Hamiltonian cycles.
+    /// It will stop as soon as it finds one valid cycle. If multiple Hamiltonian cycles exist,
+    /// only one of them will be returned.
+    ///
+    /// Returns `Some(path)` if a Hamiltonian cycle is found, where `path` is a vector
+    /// containing the indices of vertices in the cycle, starting and ending with the same vertex.
+    ///
+    /// Returns `None` if no Hamiltonian cycle exists in the graph.
+    fn find_hamiltonian_cycle(&self, start_vertex: usize) -> Option<Vec<usize>> {
+        let mut path = vec![usize::MAX; self.adjacency_matrix.len()];
+        // Start at the specified vertex
+        path[0] = start_vertex;
+
+        if self.hamiltonian_cycle_util(&mut path, 1) {
+            // To complete the cycle by returning to the starting vertex
+            path.push(start_vertex);
+            Some(path)
+        } else {
+            None
+        }
+    }
+}
+
+/// Finds a Hamiltonian cycle in a given graph represented by an adjacency matrix, if one exists, starting from a specified vertex
+pub fn find_hamiltonian_cycle(
+    adjacency_matrix: Vec<Vec<bool>>,
+    start_vertex: usize,
+) -> Result<Option<Vec<usize>>, AdjMatError> {
+    let vertices = adjacency_matrix.len();
+    // Check if the adjacency matrix is empty
+    if vertices == 0 {
+        return Err(AdjMatError::EmptyMat);
+    }
+
+    // Validate maze representation (if necessary)
+    if adjacency_matrix.iter().any(|row| row.len() != vertices) {
+        return Err(AdjMatError::ImproperMat);
+    }
+
+    // Validate start position
+    if start_vertex >= vertices {
+        return Err(AdjMatError::StartOutOfBound);
+    }
```

Comment:
What do you think about moving this _logic_ into the `new` method in `Graph`? That would require the change of its signature to:
```rust
fn new(adjacency_matrix: Vec<Vec<bool>>) -> Resut<Self, AdjMatError>
```

Reply:
I have moved this logic into appropriate methods and changed the method signatures.